crypto/fipsmodule/rand/entropy_pool.c

```diff
@@ -0,0 +1,120 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include "internal.h"
+#include "../../internal.h"
+
+#if defined(AWSLC_FIPS)
+
+// entropy_pool_validate_static_assumptions validates static assumptions on the
+// entropy pool |entropy_pool|.
+static int entropy_pool_validate_static_assumptions(
+  struct entropy_pool *entropy_pool) {
+  // By transitivity, |valid_available| and |index_read| are verified to be at
+  // most |ENTROPY_POOL_SIZE|.
+  if (entropy_pool != NULL &&
+    entropy_pool->capacity <= ENTROPY_POOL_SIZE &&
+    entropy_pool->valid_available <= entropy_pool->capacity &&
+    entropy_pool->index_read <= entropy_pool->capacity) {
+    return 1;
+  }
+  return 0;
+}
+
+// entropy_pool_ensure_can_satisfy returns 1 if the entropy pool |entropy_pool|
+// contains enough entropy to satisfy a get request of size |get_size|.
+// Returns 0 otherwise.
+static int entropy_pool_ensure_can_satisfy(struct entropy_pool *entropy_pool,
+  size_t get_size) {
+  if (entropy_pool->valid_available >= get_size) {
+    return 1;
+  }
+  return 0;
+}
+
+static int entropy_pool_consume(struct entropy_pool *entropy_pool,
+  uint8_t *get_buffer, size_t get_size) {
+
+  if ((entropy_pool->index_read + get_size) > entropy_pool->capacity) {
+    return 0;
+  }
+
+  OPENSSL_memcpy(get_buffer, entropy_pool->pool + entropy_pool->index_read,
+    get_size);
+
+  OPENSSL_cleanse(entropy_pool->pool + entropy_pool->index_read, get_size);
+  entropy_pool->index_read = entropy_pool->index_read + get_size;
+  entropy_pool->valid_available = entropy_pool->valid_available - get_size;
+
+  return 1;
+}
+
+static void entropy_pool_cannot_satisfy_request(void) {
```

Comment:
NP: to me entropy_pool_cannot_satisfy_request seems like it was getting ready to abort, it's not clear what it does.
```suggestion
static void request_external_entropy(void) {
```